### ISSUE TITLE
Imfile advances offset when output is blocked and rsyslog is restarted

### ISSUE THREAD

**Issue by StrongestNumber9** [rsyslog/rsyslog#3771]
### Expected behavior
Offset will be updated only when output is handled successfully. Would rather duplicate events than lose anything.

### Actual behavior
Offset is updated even if output is not handled, so events get lost if rsyslog gets signaled to restart.

```sh
# wc -l /logs/*.log
  21 /logs/in.log
  11 /logs/out.log

# wc -c /logs/in.log
663 /logs/in.log

# jq '' /workdir/imfile-state\:392252668
{
  "filename": "/logs/in.log",
  "prev_was_nl": 0,
  "curr_offs": 663,
  "strt_offs": 663
}

```

### Steps to reproduce the behavior
Simply write data to logs, then block output, write more data, then restart rsyslog. 

This was originally observed when relp receiver was down and source rsyslog got restarted. This behavior can be reliably reproduced by dropping rsyslog to non-privileged user and then changing file permissions to block output.

```sh
$LocalHostName test-machine
$AbortOnUncleanConfig on
$PreserveFQDN on
$PrivDropToUser rsyslog
$PrivDropToGroup rsyslog

global(
    workDirectory="/workdir"
    maxMessageSize="256k"
)

main_queue(queue.type="Direct")

module(load="imfile")

input
(
    type="imfile"
    File="/logs/in.log"
    Tag="in"
    Severity="info"
    Facility="local0"
    PersistStateInterval="1"
    ruleset="imfile"
)

template(name="onlymsg" type="list") {
    property(name="msg")
    constant(value="\n")
}

ruleset(name="imfile" queue.type="direct") {
    action(
        type="omfile"
        name="omfile"
        file="/logs/out.log"
        queue.type="direct"
        template="onlymsg"
        fileOwner="rsyslog"
        fileGroup="rsyslog"
        dirOwner="rsyslog"
        dirGroup="rsyslog"
        fileCreateMode="0600"
        dirCreateMode="0700"
    )
}
```

```sh
# ./reproduce-short.sh
  4 /logs/in.log
  3 /logs/out.log
  7 total
--
logstart
I should be in the log.
I **SHOULD** be in the log, but am not.
I should be in the log again.
--
logstart
I should be in the log.
I should be in the log again.
```

reproduce-short.sh
```sh
#!/bin/bash
mkdir /logs
chown rsyslog:rsyslog /logs
echo "logstart" > /logs/in.log
rsyslogd -i /var/run/rsyslogd.pid
echo "I should be in the log." >> /logs/in.log
sleep 2;
chown root:root /logs/out.log
kill -HUP "$(cat /var/run/rsyslogd.pid)"
sleep 2;
echo "I **SHOULD** be in the log, but am not." >> /logs/in.log
sleep 2;
kill "$(cat /var/run/rsyslogd.pid)"
chown rsyslog:rsyslog /logs/out.log
rsyslogd -i /var/run/rsyslogd.pid
sleep 2;
echo "I should be in the log again." >> /logs/in.log
sleep 2;

wc -l /logs/*.log
echo "--"
cat /logs/in.log
echo "--"
cat /logs/out.log
```

### Environment
Centos7 docker container with rsyslog from official repository
```sh
# rsyslogd -v
rsyslogd  8.1907.0 (aka 2019.07) compiled with:
        PLATFORM:                               x86_64-redhat-linux-gnu
        PLATFORM (lsb_release -d):
        FEATURE_REGEXP:                         Yes
        GSSAPI Kerberos 5 support:              Yes
        FEATURE_DEBUG (debug build, slow code): No
        32bit Atomic operations supported:      Yes
        64bit Atomic operations supported:      Yes
        memory allocator:                       system default
        Runtime Instrumentation (slow code):    No
        uuid support:                           Yes
        systemd support:                        Yes
        Config file:                            /etc/rsyslog.conf
        PID file:                               /var/run/syslogd.pid
        Number of Bits in RainerScript integers: 64

See https://www.rsyslog.com for more information.

# rsyslogd -N9
rsyslogd: version 8.1907.0, config validation run (level 9), master config /etc/rsyslog.conf
rsyslogd: End of config validation run. Bye.

# cat /etc/centos-release
CentOS Linux release 7.6.1810 (Core)

# uname -a
Linux test-machine 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by davidelang** [comment]
please enable impstats on your system

what I suspect is happening is that rsyslog is reading from the file and adding 
to the queue successfully (that is as far as imfile can see), and then when you 
restart rsyslog it's loosing the in-memory queue. you would have to enable a 
disk assisted queue and configure it to save on shutdown (and make sure your 
system timeouts give it enough time to save the queue data) to avoid loosing the 
in-memory data on restart

the pstats data will show the data read from imfile into the queue

David Lang

**Comment by StrongestNumber9** [comment]
Added impstats. Using jsonmesg template and few more rounds compared to the reproduce-short.sh 

```sh
# jq -r '."$!".metadata.fileoffset' out.log
0
12
38
64
90
116
508
539
570
601
632

# jq -r 'select(."app-name" == "rsyslogd-pstats").rawmsg' metrics.log
global: origin=dynstats
omfile: origin=core.action processed=12 failed=5 suspended=1 suspended.duration=30 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=530494 stime=1566122 maxrss=1808 minflt=398 majflt=0 inblock=0 oublock=120 nvcsw=25 nivcsw=36 openfiles=6
global: origin=dynstats
omfile: origin=core.action processed=1 failed=0 suspended=1 suspended.duration=30 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=598207 stime=1653555 maxrss=1720 minflt=396 majflt=0 inblock=0 oublock=16 nvcsw=11 nivcsw=18 openfiles=6
global: origin=dynstats
omfile: origin=core.action processed=1 failed=0 suspended=1 suspended.duration=30 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=599176 stime=1616505 maxrss=1720 minflt=395 majflt=0 inblock=0 oublock=16 nvcsw=7 nivcsw=46 openfiles=6
global: origin=dynstats
omfile: origin=core.action processed=1 failed=0 suspended=1 suspended.duration=30 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=549751 stime=1530017 maxrss=1716 minflt=397 majflt=0 inblock=0 oublock=16 nvcsw=15 nivcsw=17 openfiles=6
global: origin=dynstats
omfile: origin=core.action processed=1 failed=0 suspended=1 suspended.duration=30 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=522847 stime=1506096 maxrss=1720 minflt=394 majflt=0 inblock=0 oublock=16 nvcsw=9 nivcsw=22 openfiles=6
global: origin=dynstats
omfile: origin=core.action processed=5 failed=0 suspended=0 suspended.duration=0 resumed=0
omfile: origin=core.action processed=2 failed=0 suspended=0 suspended.duration=0 resumed=0
resource-usage: origin=impstats utime=559421 stime=1496004 maxrss=1716 minflt=383 majflt=0 inblock=0 oublock=72 nvcsw=18 nivcsw=33 openfiles=7

# wc -l *.log
   21 in.log
   24 metrics.log
   11 out.log
   56 total
```

**Comment by rgerhards** [comment]
Well, the problem here is that you are using rsyslog in a way it was not really designed for. While we try to not lose messages, we are not a transactional over-the-whole-process-of-what-we-do type of system. A core assumption is that data is put into queues, and the queues handle the data once it is placed there. Failing actions are handled by the queue and the action definition in regard to retry. We can avoid queue overrun by using proper delay mode, and so in practical deployments the problem you see never occurs (or at an extremely low probability when things already have become really bad).

Now you are running everything without queues in direct mode. We implemented direct mode in order to have an efficient way to call actions. It was never intended to be a keep-all-synchronous tool for coupling of the different components. 

I have evaluated what needs to be done to make this work in the way you have it set up. The core problem is that if we need to react to such queue errors, we need to undo everything imfile did on the way up to this point. We also need to find out where exactly the problem occured (we could lessen that requirement if we consider only direct queues for this type of handling, but still). So we would need to refactor both imfile and the file stream class to to provide entry points for

* start transaction
* rollback transaction
* commit transaction

If we do not do this, we would never get the the on-disk structures right. Even worse, this would probably complicate the components and limit it's performance, so we would possibly need to fork imfile and the stream class and provide a module for your use case. I am not convinced it is that bad, but I may be happen. In any case, I think this is a pretty decent project, not a "quick fix". It will at least take a serious amount of your support contract's engineering hours (or even overrun them).

The simple solution is to use a disk mode for the main queue.

Please let me know your thoughts. If you'd like to go for implementing transaction-like processing in imfile, we probably should coordinate via the Adiscon office on how to get you a proper quote for this body of work.

**Comment by kortemik** [comment]
```
/* Note: we currently need to return an iRet, as this is used in                                                                                                     
 * direct mode. TODO: However, it may be worth further investigating this,                                                                                           
 * as it looks like there is no ultimate consumer of this code.                                                                                                      
 * rgerhards, 2013-11-06                                                                                                                                             
 */                                                                                                                                                                  
static rsRetVal ATTR_NONNULL()                                                                                                                                       
actionCommit(action_t *__restrict__ const pThis, wti_t *__restrict__ const pWti)
```

Does this anyhow relate to this?

**Comment by rgerhards** [comment]
> Does this anyhow relate to this?

not at all

**Comment by kortemik** [comment]
Would following work?

In addition to imfile-statefile we would create a new statefile called "submitted-offset" file. This file would contain time which the imfile was instantiated (uptime stamp) plus offset of the previous offset in actual statefile. So every time actual statefile is updated, the submitted-offset file is stored with the uptime stamp and previous offset.

This would allow rsyslog to check that if the submitted-offset file is authored by the current run, if yes then just keep going as currently. If this is not however the case rsyslog would first do the normal actual imfile validation and then revert to last offset within the submitted-offset file.

This way no transactioning support needs to be in-place but we are able to resume with duplicates and therefore with certainty that everything has been sent?

I'd assume this does not necessarily require too complicated implementation compared to full transactioning support which is not the intention.

**Comment by rgerhards** [comment]
@kortemik this could work; the performance will be bad, but I assume that's of no concern for your use case. I am not sure, however, how the uptime timestamp may reliably identify the instance as new vs. existing - maybe the pid is a better choice here?

I need to check this though all layers as imfile utilizes a common component for reading (which complicates things here). Unfortunately, I am not sure if I can go through all of that this week. Will at least give it a try.

**Comment by kortemik** [comment]
@rgerhards thank you, however there is no need to push it for this week. We can't implement it before January anyways.

**Comment by kortemik** [comment]
Oh and now that I think about it, I don't see why there should be another statefile, we could just keep the information for the rollback within the current one, so it should be efficient enough?

**Comment by rgerhards** [comment]
> we could just keep the information for the rollback within the current one, so it should be efficient enough?

no, because I would need to re-read the previous state file to obtain old state. At the time the state file is written, old state is unknow.

I still do not understand when and how you would decide which of the two state files to use. Can you elaborate on this?

**Comment by kortemik** [comment]
I thought a diagram would help so I came up with this one to aid discussion


![imfile](https://user-images.githubusercontent.com/884026/70042963-7479c000-15c8-11ea-9f45-8aa4029cac2d.png)

**Comment by rgerhards** [comment]
ok, this helps. I will need to analyze how this all needs to go through the layers. I guess it's still quite some work, but now probably doable within 2 to 3 days (maybe even faster).

**Comment by rgerhards** [comment]
@kortemik Do I understand the chart correctly that you only want to persist the "fallback state file" when rsyslog starts up? e.g. if it runs for a day or so, it is ok to-resend the whole day of logs?

**Comment by kortemik** [comment]
@rgerhards I think I must have limited the case too much by the attempt to keep it small: We would rather want to persist the offset that is known that has passed through like if we have persist state interval of 1000 then we would want to store the state where the imfile was at before the next update so on startup it can seek backwards to the previous statefile update position and resend those 1000 messages if restarted.

We use persist state interval of 1 but it should work the same still.

**Comment by rgerhards** [comment]
OK, I thought this when I re-read the chart this morning. That actually might bring back the full complexity as well. I've a meeting soon, but will try to check ASAP.

**Comment by rgerhards** [comment]
I have gone once more through the code. Probably there is a very simple solution. If I understand you correctly, you do want that the state file is only updated once the full batch is successfully submitted. So rather than persisting the state file after each line read, we could only persist it after the batch has been processed.

If so, no update will be done while the file is read, only once the batch has been submitted. I can add a new parameter for this handling. The current config system does not provide a reliable way to do this. What do you think? Does that address your need?

**Comment by rgerhards** [comment]
I added a simple patch to show the concept in https://github.com/rsyslog/rsyslog/pull/4031. Maybe you can give it a try.

You need to set `persiststateinterval="0"` and `persistStateOnSubmission="on"` to get the new behaviour. Note that the current patch does NOT handle the border case where a very small file grows so that the file_id can be generated. In this case, it still persists the state file without waiting for submission. So please do not try with very small files. If the concept works for you, we can iron out that border case.

**Comment by kortemik** [comment]
Thank you! We will test this patch to find out that if it's sufficient. By
the given description it really sounds what we really need!

**Comment by rgerhards** [comment]
closed via #4031
